These scripts prepare input files for HotNet2 and run the pipeline for HotNet2

* get_cnv.py - Given a tab-separated input files with a 'name2' header for gene name and 'chrom', 'begin', 'end', generates the CNV input file for HotNet2
* get_mutation.py - Given the SNV/small indels file in Complete Genomics testvariants format with ANNOVAR annotations, generates the SNV input file for HotNet2 (genes with amino-acid changing mutations)


